src/verifier.rs

```diff
@@ -0,0 +1,406 @@
+use std::alloc::handle_alloc_error;
+
+use anyhow::Result;
+
+use crate::{
+    AffinePoint, Circuit, Curve, Field, HaloCurve, msm_execute, msm_execute_parallel,
+    msm_precompute, NUM_ROUTED_WIRES, NUM_WIRES, ProjectivePoint, Proof,
+};
+use crate::fft::{fft_precompute, ifft_with_precomputation_power_of_2};
+use crate::partition::get_subgroup_shift;
+use crate::plonk_challenger::Challenger;
+use crate::plonk_gates::evaluate_all_constraints;
+use crate::plonk_proof::OpeningSet;
+use crate::plonk_util::{eval_poly, halo_g, halo_n, pedersen_hash, powers, reduce_with_powers};
+
+const SECURITY_BITS: usize = 128;
+
+pub struct VerificationKey<C: Curve> {
+    selector_commitments: Vec<AffinePoint<C>>,
+    sigma_commitments: Vec<AffinePoint<C>>,
+    degree_log: usize,
+    degree_pow: usize,
+}
+
+pub fn verify_proof_circuit<C: HaloCurve, InnerC: HaloCurve<BaseField=C::ScalarField>>(
+    public_inputs: &[C::ScalarField],
+    proof: &Proof<C>,
+    circuit: &Circuit<C>,
+) -> Result<bool> {
+    let Proof {
+        c_wires,
+        c_plonk_z,
+        c_plonk_t,
+        o_public_inputs,
+        o_local,
+        o_right,
+        o_below,
+        halo_l,
+        halo_r,
+        halo_g,
+        schnorr_proof,
+    } = proof;
+    // Verify that the proof parameters are valid.
+    check_proof_parameters(proof);
+
+    // Check public inputs.
+    if !verify_public_inputs(public_inputs, proof) {
+        println!("Public inputs don't match.");
+        return Ok(false);
+    }
+
+    // Observe the transcript and generate the associated challenge points using Fiat-Shamir.
+    let challs = get_challenges(proof, Challenger::new(SECURITY_BITS));
+
+    let degree = circuit.degree();
+
+    let constraint_terms = evaluate_all_constraints::<C, InnerC>(
+        &proof.o_local.o_constants,
+        &proof.o_local.o_wires,
+        &proof.o_right.o_wires,
+        &proof.o_below.o_wires,
+    );
+
+    // Evaluate zeta^degree.
+    let mut zeta_power_d = challs.zeta.exp_usize(degree);
+    // Evaluate Z_H(zeta).
+    let one = <C::ScalarField as Field>::ONE;
+    let z_of_zeta = zeta_power_d - one;
```

Comment:
nit: Would prefer `zero_of_zeta` or similar to avoid confusion with the other Z

Reply:
Yes you're right, went with `zero_of_zeta`.